_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,220 @@
+</center>
+<br/>
+
+
+
+This blogpost will instruct how to run the demo locally, will describe it's components and their interactions. We will look in the details into the first pattern: Dynamic Data Partitioning.
+
+If you are only interested in learning about a specific pattern, you can jump to the respective separate section (//TODO: ? depends on break up of further blogposts):  
+ - Link 1  
+ - Link 2  
+ - Link 3  
+
+
+### Setup
+
+This blogpost is accompanied by a github repository:
+
+[](https://github.com/afedulov/fraud-detection-demo)
+
+Setup is dockerized and includes the following components:  
+
+ - Apache Kafka broker with ZooKeeper  
+ - Apache Flink [application cluster](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#flink-application-cluster)  
+ - Demo WebApp  
+
+#### Requirements:
+Demo is bundled in a self-contained package. In order to build it from sources you will need:
+
+ - git
+ - docker
+ - docker-compose
+
+ Recommended resources allocated to Docker:
+
+ - 4 CPUs
+ - 8GB RAM
+
+ You can checkout the repository and run the demo locally.
+
+#### How to run:
+
+ Execute the following commands to build the project from sources and start all required services, including Apache Flink and Apache Kafka clusters.
+
+```bash
+git clone https://github.com/afedulov/fraud-detection-demo
+docker build -t demo-fraud-webapp:latest -f webapp/webapp.Dockerfile webapp/
+docker build -t flink-job-fraud-demo:latest -f flink-job/Dockerfile flink-job/
+docker-compose -f docker-compose-local-job.yaml up
+```
+
+__Note__: Dependencies are stored in a cached Docker layer. If you later only modify the source code, not the dependencies, you can expect significantly shorter packaging times for the subsequent builds.
+
+When all components are up and running, go to `localhost:5656` in your browser.
+
+__Note__: you might need to change exposed ports in _docker-compose-local-job.yaml_ in case of collisions.
+
+The demo comes with a set of predefined rules. You can simply click the _Start_ button and after some time you should observe alerts displayed on the right side of the screen. Those alerts are the results of Flink evaluating the generated transactions stream against the predefined rules.
+
+ The sample application consists of three main parts:  
+
+ 1. Fraud Detection application (Apache Flink)  
+ 1. Frontend (React)  
+ 1. Backend (SpringBoot)  
+
+
+ Backend includes a Transactions Generator, which sends emulated transactions to Flink via Kafka. Whenever a new rule definition is created in the UI, the backend also pushes it to Flink via a separate topic. Alerts generated by Flink are consumed by the Backend and relayed to the UI via WebSockets.
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/architecture.png" width="800px" alt="Figure 2: Demo Components"/>
+ <br/>
+ <i><small>Figure 2: Demo Components</small></i>
+ </center>
+ <br/>
+
+
+### Dynamic Data Partitioning
```

Comment:
This part actually does not really talk about the core of the whole thing - how those rules are being propagated. I had to describe the rules format in order to make it clear what is this grouping key parameter and why it can change at runtime.